File: shared_resources/src/main/java/com/thewizrd/shared_resources/utils/StringUtils.java
package com.thewizrd.shared_resources.utils;

import android.text.TextUtils;

import java.util.regex.Pattern;

public class StringUtils {
    public static boolean isNullOrEmpty(String s) {
        return s == null || s.length() == 0;
    }

    public static boolean isNullOrWhitespace(String s) {
        return s == null || isWhitespace(s);
    }

    private static boolean isWhitespace(String s) {
        if (s == null)
            return true;

        for (int idx = 0; idx < s.length(); ++idx) {
            if (!Character.isWhitespace(s.toCharArray()[idx]))
                return false;
        }

        return true;
    }

    public static String toUpperCase(String s) {
        if (isNullOrEmpty(s)) {
            return s;
        }
        final char[] buffer = s.toCharArray();
        boolean capitalizeNext = true;
        for (int i = 0; i < buffer.length; i++) {
            final char ch = buffer[i];
            if (Character.isWhitespace(ch)) {
                capitalizeNext = true;
            } else if (capitalizeNext) {
                buffer[i] = Character.toTitleCase(ch);
                capitalizeNext = false;
            }
        }
        return new String(buffer);
    }

    public static String toPascalCase(String s) {
        String[] strArray = s.split("\\.", 0);
        StringBuilder sb = new StringBuilder();

        for (String str : strArray) {
            if (str.length() == 0)
                continue;

            sb.append(str.trim().substring(0, 1).toUpperCase())
                    .append(str.trim().substring(1).toLowerCase())
                    .append(". ");
        }

        return sb.toString().trim();
    }

    public static String removeNonDigitChars(CharSequence s) {
        if (TextUtils.isEmpty(s) || isNullOrWhitespace(s.toString()))
            return "";
        else {
            return s.toString().replaceAll("[^\\d.-]", "").trim();
        }
    }

    public static String removeNonDigitChars(String s) {
        if (isNullOrWhitespace(s))
            return "";
        else {
            return s.replaceAll("[^\\d.-]", "").trim();
        }
    }

    public static String removeDigitChars(String s) {
        if (isNullOrWhitespace(s))
            return "";
        else {
            return s.replaceAll("[0-9]", "").trim();
        }
    }

    public static boolean containsDigits(String s) {
        if (isNullOrWhitespace(s))
            return false;
        else {
            return Pattern.matches(".*[0-9].*", s);
        }
    }

    public static String lineSeparator() {
        return System.lineSeparator();
    }

    public static String unescapeUnicode(String s) {
        if (isNullOrWhitespace(s)) {
            return s;
        } else {
            StringBuilder sb = new StringBuilder();

            final int seqEnd = s.length();
            for (int i = 0; i < s.length(); i++) {
                // Uses -2 to ensure there is something after the &#
                char c = s.charAt(i);
                if (s.charAt(i) == '&' && i < seqEnd - 2 && s.charAt(i + 1) == '#') {
                    int start = i + 2;
                    boolean isHex = false;

                    final char firstChar = s.charAt(start);
                    if (firstChar == 'x' || firstChar == 'X') {
                        start++;
                        isHex = true;

                        if (start == seqEnd) {
                            sb.append(s.substring(i));
                            break;
                        }
                    }

                    int end = start;
                    while (end < seqEnd && s.charAt(end) != ';') {
                        end++;
                    }

                    int value;
                    try {
                        if (isHex) {
                            value = Integer.parseInt(s.substring(start, end), 16);
                        } else {
                            value = Integer.parseInt(s.substring(start, end), 10);
                        }
                    } catch (NumberFormatException nfe) {
                        sb.append(s.substring(i));
                        break;
                    }

                    char[] chars = Character.toChars(value);
                    sb.append(chars);

                    i = end;
                } else {
                    sb.append(s.charAt(i));
                }
            }

            return sb.toString();
        }
    }
}

File: shared_resources/src/main/java/com/thewizrd/shared_resources/utils/StringUtils.kt
package com.thewizrd.shared_resources.utils

import android.text.TextUtils
import java.util.regex.Pattern

object StringUtils {
    @JvmStatic
    fun String?.isNullOrEmpty(): Boolean {
        return this == null || this.isEmpty()
    }

    @JvmStatic
    fun String?.isNullOrWhitespace(): Boolean {
        return this.isNullOrBlank()
    }

    @JvmStatic
    fun String.toUpperCase(): String {
        if (this.isBlank()) {
            return this
        }
        val buffer = this.toCharArray()
        var capitalizeNext = true
        for (i in buffer.indices) {
            val ch = buffer[i]
            if (Character.isWhitespace(ch)) {
                capitalizeNext = true
            } else if (capitalizeNext) {
                buffer[i] = Character.toTitleCase(ch)
                capitalizeNext = false
            }
        }
        return String(buffer)
    }

    @JvmStatic
    fun String.toPascalCase(): String {
        val strArray = this.split("\\.".toRegex(), limit = 0)
        val sb = StringBuilder()

        for (str in strArray) {
            if (str.isEmpty()) continue

            sb.append(str.trim().substring(0, 1).uppercase())
                .append(str.trim().substring(1).lowercase())
                .append(". ")
        }

        return sb.toString().trim()
    }

    @JvmStatic
    fun CharSequence.removeNonDigitChars(): CharSequence {
        return if (TextUtils.isEmpty(this) || this.isBlank()) {
            ""
        } else {
            this.replace("[^\\d.-]".toRegex(), "").trim()
        }
    }

    @JvmStatic
    fun String.removeNonDigitChars(): String {
        return if (this.isBlank()) {
            ""
        } else {
            this.replace("[^\\d.-]".toRegex(), "").trim()
        }
    }

    @JvmStatic
    fun String.removeDigitChars(): String {
        return if (this.isBlank()) {
            ""
        } else {
            this.replace("[0-9]".toRegex(), "").trim()
        }
    }

    @JvmStatic
    fun String?.containsDigits(): Boolean {
        return if (this.isNullOrBlank()) {
            false
        } else {
            Pattern.matches(".*[0-9].*", this)
        }
    }

    @JvmStatic
    fun lineSeparator(): String {
        return System.lineSeparator()
    }

    fun String.unescapeUnicode(): String {
        if (this.isNullOrEmpty()) {
            return this
        } else {
            val sb = StringBuilder()

            val seqEnd = this.length
            var i = 0
            while (i < this.length) {
                // Uses -2 to ensure there is something after the &#
                val c = this[i]
                if (this[i] == '&' && i < seqEnd - 2 && this[i + 1] == '#') {
                    var start = i + 2
                    var isHex = false

                    val firstChar = this[start]
                    if (firstChar == 'x' || firstChar == 'X') {
                        start++
                        isHex = true
                        if (start == seqEnd) {
                            sb.append(this.substring(i))
                            break
                        }
                    }

                    var end = start
                    while (end < seqEnd && this[end] != ';') {
                        end++
                    }

                    val value = try {
                        if (isHex) {
                            this.substring(start, end).toInt(16)
                        } else {
                            this.substring(start, end).toInt(10)
                        }
                    } catch (nfe: NumberFormatException) {
                        sb.append(this.substring(i))
                        break
                    }

                    val chars = Character.toChars(value)
                    sb.append(chars)

                    i = end
                } else {
                    sb.append(this[i])
                }

                i++
            }

            return sb.toString()
        }
    }

    fun String.escapeUnicode(): String {
        // https://stackoverflow.com/a/48706264
        return native2ascii(this)
    }

    /**
     * Encode a String like äöü to \u00e4\u00f6\u00fc
     *
     * @param text
     * @return
     */
    private fun native2ascii(text: String): String {
        val sb = StringBuilder()
        for (ch in text.toCharArray()) {
            sb.append(native2ascii(ch))
        }
        return sb.toString()
    }

    /**
     * Encode a Character like ä to \u00e4
     *
     * @param ch
     * @return
     */
    private fun native2ascii(ch: Char): String {
        return if (ch > '\u007f') {
            val sb = StringBuilder()
            // write \udddd
            sb.append("\\u")
            val hex = StringBuffer(Integer.toHexString(ch.code))
            hex.reverse()
            val length = 4 - hex.length
            for (j in 0 until length) {
                hex.append('0')
            }
            for (j in 0..3) {
                sb.append(hex[3 - j])
            }
            sb.toString()
        } else {
            ch.toString()
        }
    }
}